sycl/include/sycl/vector_preview.hpp

```diff
@@ -188,212 +124,54 @@ class RoundedRangeKernel;
-    using DataType = sycl::ext::oneapi::detail::bf16::Vec##Num##StorageT;      \
-    using VectorDataType =                                                     \
-        sycl::ext::oneapi::detail::bf16::Vec##Num##StorageT;                   \
-  };
-__SYCL_DEFINE_BF16_VECSTORAGE(2)
-__SYCL_DEFINE_BF16_VECSTORAGE(3)
-__SYCL_DEFINE_BF16_VECSTORAGE(4)
-__SYCL_DEFINE_BF16_VECSTORAGE(8)
-__SYCL_DEFINE_BF16_VECSTORAGE(16)
-#undef __SYCL_DEFINE_BF16_VECSTORAGE
+using element_type_for_vector_t = typename std::conditional_t<
+#if (!defined(_HAS_STD_BYTE) || _HAS_STD_BYTE != 0)
+    std::is_same_v<T, std::byte>, std::uint8_t,
+#else
+    false, T,
+#endif
+    typename std::conditional_t<
+        std::is_same_v<T, bool>, std::int8_t,
+        typename std::conditional_t<
+            std::is_same_v<T, sycl::half>, sycl::detail::half_impl::StorageT,
+            typename std::conditional_t<
+                std::is_same_v<T, sycl::ext::oneapi::bfloat16>,
+                sycl::ext::oneapi::detail::Bfloat16StorageT, T>>>>;
 } // namespace detail
 
-template <typename T> using vec_data = detail::vec_helper<T>;
+///////////////////////// class sycl::vec /////////////////////////
+// Provides a cross-patform vector class template that works efficiently on
+// SYCL devices as well as in host C++ code.
+template <typename DataT, int NumElements> class vec {
 
-template <typename T>
-using vec_data_t = typename detail::vec_helper<T>::RetType;
+  static_assert(NumElements == 1 || NumElements == 2 || NumElements == 3 ||
+                    NumElements == 4 || NumElements == 8 || NumElements == 16,
+                "Invalid number of elements for sycl::vec: only 1, 2, 3, 4, 8 "
+                "or 16 are supported");
+  static_assert(sizeof(bool) == sizeof(int8_t), "bool size is not 1 byte");
 
-///////////////////////// class sycl::vec /////////////////////////
-/// Provides a cross-patform vector class template that works efficiently on
-/// SYCL devices as well as in host C++ code.
-///
-/// \ingroup sycl_api
-template <typename Type, int NumElements> class vec {
-  using DataT = Type;
+  static constexpr size_t AdjustedNum = (NumElements == 3) ? 4 : NumElements;
 
   // This represent type of underlying value. There should be only one field
   // in the class, so vec<float, 16> should be equal to float16 in memory.
-  using DataType = typename detail::VecStorage<DataT, NumElements>::DataType;
+  using DataType =
+      std::array<std::conditional_t<std::is_same_v<DataT, bool>, int8_t, DataT>,
+                 AdjustedNum>;
 
-  static constexpr bool IsHostHalf =
-      std::is_same_v<DataT, sycl::detail::half_impl::half> &&
-      std::is_same_v<sycl::detail::half_impl::StorageT,
-                     sycl::detail::host_half_impl::half>;
+public:
+#ifdef __SYCL_DEVICE_ONLY__
+  // Type used for passing sycl::vec to SPIRV builtins. Frontend treats
+  // ext_vector_type(1) as a scalar.
```

Comment:
I don't think it's true - https://godbolt.org/z/K7ach7Eq5

Reply:
Interesting. I guess it's the optimizer that converts operations on ext_vector_type(1) to that on scalar types, except for logical operations. Logical operations on ext_vector_type(1) results in 0 or -1, while on scalar types it results in 0 or 1.